File: cs/GenericsComparison/Factory/PrinterFactory.cs
using System;
using System.Collections.Generic;

namespace Factory
{
    public class PrinterFactory : Factory<IPrinter>
    {
        private readonly Dictionary<Type, IPrinter> _registeredPrinters = new();

        public void RegisterPrinter<T>(IPrinter factory)
        {
            _registeredPrinters.Add(typeof(T), factory);
        }
        
        public override IPrinter Build<T>()
        {
            if (_registeredPrinters.ContainsKey(typeof(T)))
            {
                return _registeredPrinters[typeof(T)];
            }
            return new ConsolePrinter();
        }
    }
}

File: java/src/examples/factory/PrinterFactory.java
package examples.factory;

import java.util.HashMap;
import java.util.Map;

public class PrinterFactory implements Factory<Printer> {
    Map<Class<?>, Printer> registeredPrinters = new HashMap<>();

    public <T> void registerFactory(Class<T> dataType, Printer printer) {
        registeredPrinters.put(dataType, printer);
    }

    @Override
    public <T> Printer build(Class<T> type) {
        Printer printer = registeredPrinters.get(type);
        if (printer != null) {
            return printer;
        }
        return new ConsolePrinter();
    }
}
